Library/Homebrew/os/mac/xcode.rb

```diff
@@ -35,28 +35,36 @@ def without_clt?
       end
 
       def prefix
-        @prefix ||=
-          begin
-            dir = MacOS.active_developer_dir
-
-            if dir.empty? || dir == CLT::MAVERICKS_PKG_PATH || !File.directory?(dir)
-              path = bundle_path
-              path.join("Contents", "Developer") if path
-            else
-              Pathname.new(dir)
-            end
-          end
+        @prefix = uncached_prefix unless instance_variable_defined?(:@prefix)
+        @prefix
+      end
+
+      def uncached_prefix
+        dir = MacOS.active_developer_dir
+        if dir.empty? || dir == CLT::MAVERICKS_PKG_PATH || !File.directory?(dir)
+          path = bundle_path
+          path/"Contents/Developer" if path
+        else
+          Pathname.new(dir)
+        end
+      end
+
+      def auto_selected?
+        !prefix.nil? && prefix.to_s != MacOS.active_developer_dir
       end
 
       def toolchain_path
         Pathname.new("#{prefix}/Toolchains/XcodeDefault.xctoolchain") if installed? && version >= "4.3"
       end
 
-      # Ask Spotlight where Xcode is. If the user didn't install the
-      # helper tools and installed Xcode in a non-conventional place, this
-      # is our only option. See: https://superuser.com/questions/390757
       def bundle_path
-        MacOS.app_with_bundle_id(V4_BUNDLE_ID, V3_BUNDLE_ID)
+        # Ask Spotlight where Xcode is if the path returned by `xcode-select` is
+        # invalid or points at the CLT. Make sure we only auto-select an Xcode
+        # installation in one of the default locations (see #default_prefix).
+        xcode_path = MacOS.mdfind(V4_BUNDLE_ID, V3_BUNDLE_ID).detect do |path|
+          "#{path}/".start_with?("/Applications/Xcode.app/", "/Developer/")
+        end
```

Comment:
I respect and understand your point of view. However, I'm not really willing to replace a major kluge with something that still feels wrong to me (i.e. I'm not convinced, predictability is more important to me than convenience). I already compromised somewhat by keeping auto-selection for what I felt were safe locations. Originally I wanted to respect the user's choice 100%, meaning always trust `xcode-select -p` and fail if Xcode is required but not selected. Do you agree that this would be the *right thing* to do, if compatibility with previous behavior wasn't a concern?

Maybe it would be a good idea to print in `brew config` if the listed Xcode was auto-selected or actually selected by the user via `xcode-select` (as shown in first comment), irrespective of the outcome of the remaining discussion?

> Even excluding `/Volumes` may break things for people with small SSDs who stick Xcode on another partition for space reasons.

I'm not sold on this. Those people can use `xcode-select -s` to select the Xcode the have on an external disk and everything will be fine. I've been using Homebrew for a few years now and *always* had Xcode selected (never the CLT) and can't remember having issues due to this. Of course I've been using up-to-date Xcode and OS X releases most of the time, so that might have contributed to this relatively smooth experience.